core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -913,9 +915,18 @@ private[kafka] class Processor(val id: Int,
               } else {
                 val connectionId = receive.source
                 val context = new RequestContext(header, connectionId, channel.socketAddress,
-                  channel.principal, listenerName, securityProtocol)
+                  channel.principal, listenerName, securityProtocol,
+                  channel.clientInformation.orElse(ClientInformation.EMPTY))
                 val req = new RequestChannel.Request(processor = id, context = context,
                   startTimeNanos = nowNanos, memoryPool, receive.payload, requestChannel.metrics)
+                if (header.apiKey == ApiKeys.API_VERSIONS) {
+                  val apiVersionsRequest = req.body[ApiVersionsRequest]
+                  if (apiVersionsRequest.isValid) {
+                    selector.updateClientInformation(channel, new ClientInformation(
+                      apiVersionsRequest.data.clientSoftwareName,
+                      apiVersionsRequest.data.clientSoftwareVersion))
+                  }
+                }
```

Comment:
I have chosen to intercept the AVR request here to update the metric because it would be overly complicated to wire things up to the API layer. The parsing happens in the processor thread anyway so the impact is low.